### DISCUSSION TITLE
Setting up an upstream branch ( where and how )

### DISCUSSION THREAD

**Discussion by guennt** [community/community#46359] [New to GitHub]
Hello, 

I have installed an Obsidian PlugIn that makes a backup for my dataset. 

For this I have set up a REPOSITORY under GitHub.com. 

But the Developer Console shows alternately always 2 errors: 

1. NO UPSTREAM BRANCH 

2. CAN NOT RUN COMMAND 

The most important seems to me to be point 1 ). 

I don't know where and how to set up a branch. Does it have to happen in the Obsidian PlugIn or on GitHub. If the latter is the case, where and how do I have to do it. 

I hope I can get support.
Many thanks
Guenter 

<img width="558" alt="#1-can not run " src="https://user-images.githubusercontent.com/121873100/216776383-06c3d6c0-0231-4f1d-963e-eca5d4a92cb5.png">

<img width="614" alt="#2 no upstream branch " src="https://user-images.githubusercontent.com/121873100/216776402-514fb198-c2aa-4f08-9be1-3598dccb6ab4.png">

**Comment by shayan1234554321** [comment]
🍊The error message "No upstream branch" typically indicates that the Git repository you are using for your Obsidian backup is not set up correctly. To resolve this issue 🤝 , you need to create an upstream branch in the repository and configure the Obsidian plugin to use it.

Here's how you can do it:

1. On GitHub, create a new branch 👉 in your repository and give it a descriptive name, such as "backup".
2. In Obsidian, open the settings for the backup plugin you installed and look for an option to specify the upstream branch. This option may be located in a section called "Remote repository", "Git settings", or something similar.
3. Enter the name of the branch you created in GitHub (e.g. "backup") as the upstream branch in the plugin settings.
4. Save 💪 the changes to the plugin settings and try running the backup again to see if the error has been resolved.

🍊The error message "Can not run command" is likely related to an issue with the backup plugin you installed in Obsidian. To resolve this issue, you can try the following steps:

1. Check the plugin documentation or support resources: Review the documentation or support resources for the backup plugin to see if there are any known issues or requirements for using the plugin ⚡️.
2. Verify plugin installation: Ensure that the plugin is installed correctly and that all dependencies are met. Some plugins may require additional setup or configuration, so be sure to review the plugin instructions carefully.
3. Check for conflicts 👊🏼 : If you have multiple plugins installed in Obsidian, there may be a conflict between the backup plugin and another plugin. Try disabling other plugins to see if the issue is resolved.
4. Reinstall the plugin: If the previous steps do not resolve the issue, try uninstalling and reinstalling the plugin.

🤳🏽Contact plugin support: If you are still encountering the "Can not run command" error after trying these steps, you can reach out to the plugin developer or the Obsidian support team for further assistance.
🤳🏽If you are still encountering the "No upstream branch" error after following these steps, you can try reaching out to the plugin developer or to the Obsidian support team for further assistance.

Note ✍🏽 : Obsidian does not have a built-in option for setting up a branch, but some backup plugins may require you to set up a branch in your repository in order to properly back up your data. If this is the case, the plugin documentation should explain how to do this.

**Comment by guennt** [comment]
Hi Shayan, 
thank you for the very detailed advice on how to solve my problems. 

First I would like to solve the first point " no upstream branch". 

The explanations are clear and well understandable. 
But what I do not manage, despite intensive search, where I can find in the respository the place where I can enter the branch. 
This should be simple, but is apparently totally hidden. Where exactly can I do that please. 
Therefore I would be very grateful for any further support. 

Many thanks for your help.
Guenter 

PS: I have gone through all menu items but without success. 

<img width="754" alt="Bildschirmfoto 2023-02-18 um 16 33 31" src="https://user-images.githubusercontent.com/121873100/219874952-a99283ab-b748-4d14-a0bd-0cd9def362e9.png">

**Reply by shayan1234554321** [reply]
You have to write that in the terminal. 
If you do not have an upstream branch. Then whichever branch you create and push the first time would be the upstream branch ( main branch ).
And if you want to create a main branch by itself, try adding a readme.md file while creating a new repository. A new branch named "main" would be the upstream branch.

To make branch, try these
git checkout -b "branch-name"  //It will create a new branch
git push origin branch-name // It will push your branch to the github .. From where you can pull it

Maybe it helps 🥰

**Comment by guennt** [comment]
Hello  shaman, 

I am glad that I got another answer from you. I have also tried to implement this. The command to create a branch is always rejected. I also tried with similar commands I found on the internet. But in no case did the creation of a branch work. 

What else should I do. It can't be that difficult. 

Many thanks for your further response.
Guenter 

<img width="568" alt="Bildschirmfoto 2023-02-21 um 17 07 58" src="https://user-images.githubusercontent.com/121873100/220399876-5f122391-06ce-498e-86a3-93eefc581f01.png">

**Reply by shayan1234554321** [reply]
The "not a git repository" occurs when you are not in a folder that is initialized with github.
I think the problem with you is that you are not able to clone your repository 🤐 .
- Create a repository
- Copy the link as in screenshot 
![image](https://user-images.githubusercontent.com/18634639/220420866-afa9ee09-9fb6-44ac-97bd-f2863e0cec40.png)
- Go to your desktop in terminal
- Write "git clone then-the-link"
- now cd your-folder-that-just-created 🤪
- Now you can try to make a branch

Even if then the problem occur, we can have a zoom call ( I am also not an expert by the way 🥶 )
You can contact me via email kshayan091@gmail.com
Or linkedin https://www.linkedin.com/in/shayan-khan20/ 🤗

**Reply by guennt** [reply]
Hello shayan,

I was very pleased to receive your reply again. This is really great. I wouldn't have gotten to this point without you. 

It seems that this is the right way. Unfortunately, it fails at a simple point, namely the password entry for my repositorie. Although I know my password, no input can be made in the terminal. It is displayed as locked with the lock symbol. ( pls screenshot ) 

Thereupon I researched and found the following at this point https://docs.github.com/en/get-started/getting-started-with-git/why-is-git-always-asking-for-my-password.  
"When Git prompts you for your password, enter your personal access token. Alternatively, you can use a credential helper like Git Credential Manager. 

**Password-based authentication for Git has been removed in favor of more secure authentication methods.** For more information, see "Creating a personal access token.“

Personal access tokens are an alternative to using passwords for authentication to GitHub when using the GitHub API or the command line.

On this page one can find a macOS package for the Git Credential Manager. 
https://github.com/GitCredentialManager/git-credential-manager/blob/release/docs/install.md

The problem: I have no idea **which package is the right one for my constellation**. Can you say something about this. My macOS operating system is "MAC OS Monterey Version 12.6. 

Many thanks for your further support. 
Guenter 

PS:
It is already made quite difficult to get basically simple things up and running. 
<img width="583" alt="#1-sha" src="https://user-images.githubusercontent.com/121873100/221168038-2313fff7-fee1-4ebc-a9d5-6e03abbef299.png">

<img width="1192" alt="#2-sha" src="https://user-images.githubusercontent.com/121873100/221168107-cfda28fc-b0b1-4dff-8bd8-8ca3f56e5fcc.png">

**Reply by guennt** [reply]
Hi Shayan,

I have not been idle and have been looking for a solution in the meantime. According to what I had read, I created a personal token, but unfortunately it did not bring success either. 
What is all this complicated !!!

Guenter 
Pls screenshots

<img width="581" alt="# 2023-02-24 " src="https://user-images.githubusercontent.com/121873100/221203331-00772866-e64e-4860-8033-0aaa8fb83138.png">


<img width="890" alt="token 2023-02-24 um 15 16" src="https://user-images.githubusercontent.com/121873100/221203418-5b92cd8d-8e97-487d-89be-895733280703.png">

**Reply by guennt** [reply]
Hi shayan,

Some more changes I made in the meantime: 

I have updated the TOKEN via Keychain Access as shown in the screenshot. 
I replaced the old password with the token. 
Also this time the calculation does not work. 
Also this input is blocked by the key symbol. 

**Now my ideas how it could work are really at the end**. All because of this strange key symbol that can't be eliminated. 

What you can see below, I noticed this on the help page. There is the following procedure described:

Once you have a token, you can enter it instead of your password when performing Git operations over HTTPS.

For example, on the command line you would enter the following:

$ git clone https://github.com/USERNAME/REPO.git
Username: YOUR_USERNAME
**Password: YOUR_TOKEN**

I think REPO is the abbreviation for repository.

In this respect my data meets these requirements exactly
$ git clone https://github.com/guennt/myGitHub-Ordner.git

Why all this still does not work is completely incomprehensible to me 

That's it I want to add, so your information is complete.

Many thanks for your response.
Guenter 

<img width="542" alt="#2-2023-02-24 um 20 04" src="https://user-images.githubusercontent.com/121873100/221274990-ac731b31-7ca5-4119-a791-ab833b16c964.png">

**Reply by shayan1234554321** [reply]
As I am not a Mac user but I will still try to help you. 🥵
It sounds like you are trying to enter your GitHub password in the terminal on your macOS Monterey system, but the input is not registered 🥲 . One possibility is that the password is hidden as you type, which is a common security measure. To check if this is the case, try typing your password into the terminal and then pressing Enter. Even though you can't see what you're typing, the characters should still be registering 🤩 .  

If that doesn't work, another possibility is that the terminal is in "password" or "secure input" mode. This mode prevents any input from being displayed on the screen, including passwords. To turn off password mode, you can try pressing the Control key and then typing "v". This should turn off password mode and allow you to enter your password normally.

Based on the file names and operating system information you provided, it looks like you are using macOS Monterey Version 12.6, which is an ARM-based system. In that case, you should choose one of the following files:

    gcm-osx-arm64-2.0.886.pkg
    gcm-osx-arm64-2.0.886.tar.gz

The first file is an installer package for macOS, while the second file is a compressed archive that you can extract and use without installing. Both files are for ARM-based systems like yours. 🥶 

As I mentioned that I am not a Mac user, I am getting all of those answers from chatGPT 💻 . Please try it and tell it whatever you have to say, try pasting some errors and give an explanation, It will help you a lot.
Here are the screenshots by which I got the help 🌻 .

![Screenshot 2023-02-25 at 01-27-52 ChatGPT](https://user-images.githubusercontent.com/18634639/221285211-48849c42-12da-463a-be16-3c870e75940f.png)
![Screenshot 2023-02-25 at 01-28-13 ChatGPT](https://user-images.githubusercontent.com/18634639/221285225-6a9c6fec-be58-4c89-8c2a-cbb499f2f465.png)

**Reply by guennt** [reply]
Hello Shayan, 

Thanks for the quick reply with the 3 suggestions:

No. 1 )  I also can not make an invisible input. The input is completely blocked. 

No. 2) the key combination Control "CTRL" + "V" also shows no effect. 

No. 3) the installation is aborted with an error message. " cannot be installed on this computer " ( pls screenshot ) 

It really seems everything is highly complex and obscure. 

**How to solve the new problem ?** 

Many thanks for your help.
Guenter 

<img width="1193" alt="#1-xx" src="https://user-images.githubusercontent.com/121873100/221369137-51d14e80-ca5c-4eb7-aa6d-6cf9f6a04fc9.png">

**Reply by shayan1234554321** [reply]
I am so sorry but I tried to help you with all of the information I got. That was all I could do, sorry

**Reply by guennt** [reply]
Hi shayan,
Yes, that was very constructive of you.
I would also like to thank you for that.
It's a real pity that this issue is so difficult to solve. 

Many thanks
Guenter
